File: js/src/model/conditions/AndCompositeRule.js
import CompositeRule from './CompositeRule'
/**
 * 以"并且"为关系组织的条件规则组合
 */
class AndCompositeRule extends CompositeRule{
    /**
     * 以"并且"为关系组织的条件规则组合
     * @param {(SimplexRule|SameRangeRule|AndCompositeRule|OrCompositeRule)[]} components
     */
    constructor(components) {
        super(components);
        this.combinator = "&";
    }

    combineResult(b1, b2) {
        return b1&&b2;
    }

    /**
     * 检查items是否满足当前规则
     * @param {Item []} items
     * @returns {boolean}
     */
    check(items) {
        for(let rule of this.components) {
            if(rule) {
                let r = rule.check(items);
                if (!r) {
                    return false;
                }
            }
        }
        return true;
    }
}

export default AndCompositeRule;

File: java/src/main/java/com/github/thinhunan/wonder8/promotion/rule/model/AndCompositeRule.java
package com.github.thinhunan.wonder8.promotion.rule.model;


import com.github.thinhunan.wonder8.promotion.rule.model.validate.RuleValidateResult;

import java.util.ArrayList;
import java.util.List;

/**
 * 以"并且"为关系组织的条件规则组合
 */
public class AndCompositeRule extends CompositeRule {

    public AndCompositeRule(List<RuleComponent> components){
        this.components = components;
    }


    public AndCompositeRule(){
        this.components = new ArrayList<>();
    }
    @Override
    String getCombinator() {
        return "&";
    }

    @Override
    boolean CombineResult(boolean r1, boolean r2) {
        return (r1&&r2);
    }


    @Override
    public boolean check(List<Item> items) {
        for(RuleComponent rule : components) {
            RuleValidateResult r = rule.validate(items);
            if(!r.isValid()){
                return false;
            }
        }
        return true;
    }
}
